setuptools/tests/integration/test_pip_install_sdist.py

```diff
@@ -0,0 +1,272 @@
+"""Integration tests for setuptools that focus on building packages via pip.
+
+The idea behind these tests is not to exhaustively check all the possible
+combinations of packages, operating systems, supporting libraries, etc, but
+rather check a limited number of popular packages and how they interact with
+the exposed public API. This way if any change in API is introduced, we hope to
+identify backward compatibility problems before publishing a release.
+
+The number of tested packages is purposefully kept small, to minimise duration
+and the associated maintenance cost (changes in the way these packages define
+their build process may require changes in the tests).
+"""
+import importlib
+import json
+import os
+import subprocess
+import sys
+import tarfile
+from enum import Enum
+from glob import glob
+from hashlib import md5
+from itertools import chain
+from urllib.request import urlopen
+from zipfile import ZipFile
+
+import pytest
+import tomli as toml
+from packaging.requirements import Requirement
+
+
+pytestmark = pytest.mark.integration
+
+
+LATEST, = list(Enum("v", "LATEST"))
+"""Default version to be checked"""
+# There are positive and negative aspects of checking the latest version of the
+# packages.
+# The main positive aspect is that the latest version might have already
+# removed the use of APIs deprecated in previous releases of setuptools.
+
+
+# Packages to be tested:
+# (Please notice the test environment cannot support EVERY library required for
+# compiling binary extensions. In Ubuntu/Debian nomenclature, we only assume
+# that `build-essential`, `gfortran` and `libopenblas-dev` are installed,
+# due to their relevance to the numerical/scientific programming ecosystem)
+EXAMPLES = [
+    ("numpy", LATEST),  # custom distutils-based commands
+    ("pandas", LATEST),  # cython + custom build_ext
+    ("sphinx", LATEST),  # custom setup.py
+    ("pip", LATEST),  # just in case...
+    ("pytest", LATEST),  # uses setuptools_scm
+    ("mypy", LATEST),  # custom build_py + ext_modules
+
+    # --- Popular packages: https://hugovk.github.io/top-pypi-packages/ ---
+    ("botocore", LATEST),
+    ("kiwisolver", "1.3.2"),  # build_ext, version pinned due to setup_requires
+    ("brotli", LATEST),  # not in the list but used by urllib3
+]
+
+
+# Some packages have "optional" dependencies that modify their build behaviour
+# and are not listed in pyproject.toml, others still use `setup_requires`
+EXTRA_BUILD_DEPS = {
+    "sphinx": ("babel>=1.3",),
+    "kiwisolver": ("cppy>=1.1.0",)
+}
+
+
+# By default, pip will try to build packages in isolation (PEP 517), which
+# means it will download the previous stable version of setuptools.
+# `pip` flags can avoid that (the version of setuptools under test
+# should be the one to be used)
+PIP = (sys.executable, "-m", "pip")
+SDIST_OPTIONS = (
+    "--ignore-installed",
+    "--no-build-isolation",
```

Comment:
I'd prefer to use the build-isolation, as that's the most common scenario. Is there another option that can be passed that means "if requested, get setuptools from the current directory"?

Reply:
I asked about this possibility in the pip discord, and based on the answers there, my conclusion is that if we want to force a specific version of setuptools we have to keep the `--no-build-isolation` flag, but instead create a separated virtualenv and install the build deps there, to simulate pip's isolation feature.

---
An alternative suggestion was to try `PIP_CONSTRAINTS`.
To test this suggestion I locally tried:

- first build a wheel of the development version of setuptools via `python -m build --wheel`
  (`PIP_CONSTRAINTS` do not accept editable installs)
- Create a `constraints.txt` file with:
  ```
  setuptools @ file:///home/<...>/setuptools/dist/setuptools-58.5.3.post20211113-py3-none-any.whl
  ```
- Running:
1. `DISTUTILS_DEBUG=1 pip --verbose install -c constraints.txt dist/Brotli-1.0.9.zip`
2. `DISTUTILS_DEBUG=1 pip --verbose install -c constraints.txt --use-pep517 dist/Brotli-1.0.9.zip`
3. `DISTUTILS_DEBUG=1 PIP_CONSTRAINTS=./constraints.txt pip --verbose install --use-pep517 dist/Brotli-1.0.9.zip`

For (1) pip does not seem to respect the isolation very much, also there is nothing in the stdout/stderr that could be used to assert during the tests that the correct version of setuptools is being used.

For both (2) and (3), pip ignores the constraint and re-downloads `setuptools-58.5.3-py3-none-any.whl` from PyPI.

Reply:
The reason why I went with `pip install -t` instead of creating a virtual environment was purely to avoid the overhead and try to speed up the integration tests a little bit.

With that in mind, something I can do is to create a virtualenv for each test case and install the current version of setuptools under development in it. This way we can simulate pip's isolation, despite using the `--no-build-isolation` flag.

That will add a bit of overhead.